torchvision/prototype/models/_utils.py

```diff
@@ -1,13 +1,79 @@
 
-from ._api import WeightsEnum
+from torch import nn
+from torchvision.prototype.utils._internal import kwonly_to_pos_or_kw
 
+from ._api import WeightsEnum
 
 W = TypeVar("W", bound=WeightsEnum)
+M = TypeVar("M", bound=nn.Module)
 V = TypeVar("V")
 
 
+def handle_legacy_interface(**weights: Tuple[str, Union[Optional[W], Callable[[Any, Dict[str, Any]], Optional[W]]]]):
+    """Decorates a model builder with the new interface to make it compatible with the old.
+
+    In particular this handles two things:
+
+    1. Allows positional parameters again, but emits a deprecation warning in case they are used. See
+        :func:`torchvision.prototype.utils._internal.kwonly_to_pos_or_kw` for details.
+    2. Handles the default value change from ``pretrained=False`` to ``weights=None`` and ``pretrained=True`` to
+        ``weights=Weights`` and emits a deprecation warning with instructions for the new interface.
+
+    Args:
+        **weights (Tuple[str, Union[Optional[W], Callable[[Any, Dict[str, Any]], Optional[W]]]]): Deprecated parameter
+            name and default value for the legacy ``pretrained=True``. The default value can be a callable in which
+            case it will be called with the value of ``pretrained`` and the other keyword arguments. For example:
+    """
+
+    def outer_wrapper(builder: Callable[..., M]) -> Callable[..., M]:
+        @kwonly_to_pos_or_kw
+        @functools.wraps(builder)
+        def inner_wrapper(*args: Any, **kwargs: Any) -> M:
+            for weights_param, (pretrained_param, default) in weights.items():  # type: ignore[union-attr]
+                # If neither the weights nor the pretrained parameter as passed, or the weights argument already use
+                # the new style (not a boolean like pretrained did), there is nothing to do.
+                if (weights_param not in kwargs and pretrained_param not in kwargs) or not isinstance(
+                    kwargs.get(weights_param, False), bool
+                ):
+                    continue
+
+                # If the pretrained parameter was passed as positional argument, it is now mapped to
+                # `kwargs[weights_param]`. This happens because the @kwonly_to_pos_or_kw use the current signature to
+                # infer the names of positionally passed arguments and thus has no knowledge that there used to be a
+                # pretrained parameter.
+                pretrained_positional = weights_param in kwargs
+                pretrained_arg = kwargs.pop(weights_param if pretrained_positional else pretrained_param)
+                if pretrained_arg:
+                    default_weights_arg = default(pretrained_arg, kwargs) if callable(default) else default
+                    if not isinstance(default_weights_arg, WeightsEnum):
+                        raise ValueError(f"No weights available for model {builder.__name__}")
+                else:
+                    default_weights_arg = None
```

Comment:
I thinks this https://github.com/pytorch/vision/pull/5001#discussion_r760984921 remains unaddressed. You have renamed the parameter everywhere. See if you agree with my point on the renaming and happy to discuss this if you dont.

Reply:
I agree in the old implementation this was confusing, but I've refactored a little different. Instead of checking if `default_weights_arg is None`, we now simply enforce that it should be a `Weights` instance. Through this we now have `default_weights_arg: Optional[Weights]`, which hopefully is not confusing anymore. Thoughts?

Reply:
OK you sold it. Sounds good to me :)